slo-monitor/src/monitors/pod_monitor.go

```diff
@@ -0,0 +1,356 @@
+
+	eventSelector := fields.Set{
+		"involvedObject.kind": "Pod",
+		// this is actually source.Component
+		"source": "kubelet",
+	}.AsSelector().String()
+
+	eventController := NewWatcherWithHandler(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.FieldSelector = eventSelector
+				obj, err := pm.kubeClient.Core().Events(v1.NamespaceAll).List(options)
+				return runtime.Object(obj), err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = eventSelector
+				return pm.kubeClient.Core().Events(v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Event{},
+		func(obj interface{}) error {
+			e, ok := obj.(*v1.Event)
+			if !ok {
+				glog.Errorf("Failed to cast observed object to *v1.Event: %v", obj)
+				return nil
+			}
+			pm.handleEventUpdate(e)
+			return nil
+		},
+		func(obj interface{}) error {
+			return nil
+		},
+	)
+	go eventController.Run(stopCh)
+	go wait.Until(pm.purgeOutdated, 15*time.Second, stopCh)
+
+	select {
+	case <-stopCh:
+		glog.Infof("Stop channel closed. Performance SLO monitor is exiting.")
+	}
+}
+
+func (pm *PodStartupLatencyDataMonitor) purgeOutdated() {
+	now := time.Now()
+	pm.Lock()
+	defer pm.Unlock()
+	if len(pm.toDelete) == 0 {
+		return
+	}
+	left := []string{}
+	for _, podKey := range pm.toDelete {
+		if data, ok := pm.PodStartupData[podKey]; ok {
+			if data.deletedTimestamp.Add(purgeAfter).After(now) {
+				left = append(left, podKey)
+				continue
+			}
+			delete(pm.PodStartupData, podKey)
+		}
+	}
+
+	pm.toDelete = left
+}
+
+func (pm *PodStartupLatencyDataMonitor) handlePodUpdate(p *v1.Pod) {
+	rv, err := storage.ParseListResourceVersion(p.ResourceVersion)
+	if err != nil {
+		glog.Errorf("Failed to parse resourceVersion %v from Pod %v. Skipping.", p.ResourceVersion, getPodKey(p))
+		return
+	}
+	if rv <= pm.startPodRV {
+		// still processing backlog from before startup
+		pm.markAsAlreadyRunning(getPodKey(p))
+		return
+	}
+	if isRunning, create, running := checkPodAndGetStartupLatency(p); isRunning {
+		go pm.insertPodRunningTime(getPodKey(p), create, running)
+	}
+}
+
+func (pm *PodStartupLatencyDataMonitor) handlePodDelete(p *v1.Pod) {
+	pm.Lock()
+	defer pm.Unlock()
+
+	key := getPodKey(p)
+	data, ok := pm.PodStartupData[key]
+	if !ok {
+		data = PodStartupMilestones{}
+		data.startedPulling = time.Unix(0, 0)
+	}
+	data.deletedTimestamp = time.Now()
+	pm.PodStartupData[key] = data
+	pm.toDelete = append(pm.toDelete, key)
+}
+
+func (pm *PodStartupLatencyDataMonitor) handleEventUpdate(e *v1.Event) {
+	rv, err := storage.ParseListResourceVersion(e.ResourceVersion)
+	if err != nil {
+		glog.Errorf("Failed to parse resourceVersion %v from Event %#v. Skipping.", e.ResourceVersion, e)
+		return
+	}
+
+	ref := &e.InvolvedObject
+	if err != nil {
+		glog.Errorf("Failed to cast involvedObject onto ObjectReference. Skipping.", e.ResourceVersion, e)
+		return
+	}
+	if ref.Kind != "Pod" {
```

Comment:
We do, this is just in case something would break the watch. I guess it's unnecessary.